### PR TITLE
Fix allocempty

### PR DESCRIPTION
**Pull request by abergeron**
This makes the custom values_eq_approx for DebugMode work.

### CODE REVIEW

File: theano/compile/debugmode.py [hunk 1]
```diff
@@ -945,7 +952,10 @@ def _find_bad_optimizations0(order, reasons, r_vals):
                 r_val = r_vals[r]
                 assert r.type == new_r.type
 
-                if hasattr(new_r, 'values_eq_approx'):
+                if hasattr(new_r.tag, 'values_eq_approx'):
```
Comment (lamblin) [inline theano/compile/debugmode.py]: Why put it in `tag` rather than in the variable itself?

Comment (abergeron) [inline theano/compile/debugmode.py]: Because cloning doesn't keep random attributes you graft on variables.  But it does preserve the contents of tag.

**Comment by abergeron** [comment]
Sorry, I tackled on a somewhat unrelated commit here.  It fixes a problem where constant of CDataType would get merged but they contained different data.  This prevent merging at all, but I think is the best since we can't actually compare the contained data.

File: theano/compile/debugmode.py (hunk 1 above)
Comment (lamblin) [inline theano/compile/debugmode.py]: OK, makes sense.

**Comment by nouiz** [comment]
This seem ok for me. Can you add a test that would trigger the bad merge? The graph can contain just 2 of the conv descriptor op and nothing else.

**Comment by lamblin** [comment]
WFM as well.

**Pull request merged**